### PR TITLE
First implementation of library

### PR DESCRIPTION
**Pull request by samstickland**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+    steps:
+      - uses: actions/checkout@v3
+
+      - uses: ruby/setup-ruby@v1
+        with:
+          bundler-cache: true
+
+      - name: Tests
+        run: bundle exec ruby test/automated.rb
```
Comment (arepnikov) [inline .github/workflows/ci.yml]:
```suggestion
        run: bundle exec ruby test/automated.rb

```

File: test/automated.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+require_relative "./test_init"
+
+TestBench::CLI.(
+  exclude_file_pattern: %r{/_|sketch|(_init\.rb|_tests\.rb)\z}
+)
```
Comment (AlfonsoUceda) [inline test/automated.rb]: TestBench::Run.('test/automated')

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+name: CI
+
+on: push
+
+jobs:
+  build:
+    runs-on: self-hosted
```
Comment (AlfonsoUceda) [inline .github/workflows/ci.yml]:
```suggestion
    runs-on: ubuntu-latest
```

File: lib/log_ignored_message/controls/id.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+module Log-ignored-messageComponent
```
Comment (arepnikov) [inline lib/log_ignored_message/controls/id.rb]:
```suggestion
module LogIgnoredMessage
```

File: lib/log_ignored_message/log_ignored.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    extend Configure::Macro
+    include Log::Dependency
+    include Dependency
+
+    def self.inherited(cls)
+      cls.class_exec do
+        dependency_module = Module.new do
+          define_singleton_method :included do |reciever_class|
+            reciever_class.class_exec do
+              ::Dependency::Attribute.define(self, :log_ingored, cls)
```
Comment (arepnikov) [inline lib/log_ignored_message/log_ignored.rb:12]:
```suggestion
              ::Dependency::Attribute.define(self, :log_ignored, cls)
```

File: lib/log_ignored_message/log_ignored.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
+module LogIgnoredMessage
+  class LogIgnored
+    extend Configure::Macro
+    include Log::Dependency
+    include Dependency
+
+    def self.inherited(cls)
+      cls.class_exec do
+        dependency_module = Module.new do
+          define_singleton_method :included do |reciever_class|
```
Comment (arepnikov) [inline lib/log_ignored_message/log_ignored.rb:10]:
```suggestion
          define_singleton_method :included do |receiver_class|
```

File: lib/log_ignored_message/log_ignored.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            reciever_class.class_exec do
+              ::Dependency::Attribute.define(self, :log_ingored, cls)
+
+              define_method :log_ingored do
+                @log_ingored ||= cls.configure self
+              end
```
Comment (arepnikov) [inline lib/log_ignored_message/log_ignored.rb]:
```suggestion
          define_singleton_method :included do |receiver_class|
            receiver_class.class_exec do
              ::Dependency::Attribute.define(self, :log_ignored, cls)

              define_method :log_ignored do
                @log_ignored ||= cls.configure self
              end
```

File: lib/log_ignored_message/log_ignored.rb [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+          end
+        end
+
+        const_set :Dependency, dependency_module
+      end
+    end
+
+    self.inherited(self)
+
+    configure :log_ingored_message
```
Comment (arepnikov) [inline lib/log_ignored_message/log_ignored.rb]:
```suggestion
    configure :log_ignored_message
```

File: test/automated/log_ignored/substitute.rb [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+    context "When not invoked" do
+      test "called? returns false" do
+        refute substitute.called?
+      end
+    end
+
+    context "When invoked" do
+      substitute.(message, reason)
+    end
```
Comment (arepnikov) [inline test/automated/log_ignored/substitute.rb]: this context should include all below tests

File: test/automated/log_ignored/substitute.rb [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+    test "called? returns false" do
```
Comment (arepnikov) [inline test/automated/log_ignored/substitute.rb]:
```suggestion
    test "called? returns true" do
```

File: test/automated/log_ignored/substitute.rb [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+      assert substitute.called?
+    end
+
+    test "called_with? returns true for maatching arguments" do
```
Comment (arepnikov) [inline test/automated/log_ignored/substitute.rb]:
```suggestion
    test "called_with? returns true for matching arguments" do
```

File: test/automated/log_ignored/substitute.rb [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+      refute substitute.called_with?(message, other_reason)
+    end
+
+    test "called_with? returns false for non matching message" do
+      other_message = Controls::Message::Other.example
+
+      refute substitute.called_with?(other_message, reason)
+    end
+  end
```
Comment (arepnikov) [inline test/automated/log_ignored/substitute.rb]: remove it - it is a copy of above test

**Review by arepnikov — CHANGES_REQUESTED** [review]
few typos

File: lib/log_ignored_message/log_ignored.rb [hunk 13]
```diff
@@ -0,0 +1,62 @@
+module LogIgnoredMessage
+  class LogIgnored
```
Comment (arepnikov) [inline lib/log_ignored_message/log_ignored.rb]: maybe we can add it to eventide namespace:

```suggestion
class Log
  class IgnoredMessage
```
but you need to require to load Log namespace first and dependency could be: `:log_ignored_message`

Comment (arepnikov) [reply lib/log_ignored_message/log_ignored.rb]: and to be honest I think it should still be a logger dependency, which has method `ignore_message`

because we could need `logger.info` in some components. We might need to log some other info than ignored message, so it will require to include second logger into class, what could be weird.

It dependency could inherit from Log or delegate at some of methods, like:
:fatal, :error, :warn, :info, :debug, :trace, which are default log levels

Comment (arepnikov) [reply lib/log_ignored_message/log_ignored.rb]: Oh!  we even could create our own log level: `ignore_message` which could have his own formatter. his level could be equal to info level.

And if we will change the way we include logger in our class it should be possible to use eventide substitute without tag: :ignore, which might allow test it without extra effort/configuration in tests

It will require some investigation, but I can do it if you are interested

File: lib/log_ignored_message.rb [hunk 14]
```diff
@@ -0,0 +1,4 @@
+require 'configure'
```
Comment (AlfonsoUceda) [inline lib/log_ignored_message.rb:1]: I think you need to do `Configure.activate`

File: lib/log_ignored_message/log_ignored.rb [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    extend Configure::Macro
+    include Log::Dependency
+    include Dependency
+
+    def self.inherited(cls)
```
Comment (AlfonsoUceda) [inline lib/log_ignored_message/log_ignored.rb]: Why is this needed?

File: test/automated/log_ignored/substitute.rb (hunk 10 above)
Comment (AlfonsoUceda) [reply test/automated/log_ignored/substitute.rb]: I would use this
```suggestion
    test "Is called" do
```

File: test/automated/log_ignored/substitute.rb [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+context "LogIgnored" do
+  context "Substitute" do
+    message = Controls::Message.example
+    reason = Controls::Reason.example
+
+    substitute = LogIgnored::Substitute.build
+
+    context "When not invoked" do
+      test "called? returns false" do
```
Comment (AlfonsoUceda) [inline test/automated/log_ignored/substitute.rb]:
```suggestion
      test "Is not called" do
```

File: test/automated/log_ignored/substitute.rb (hunk 11 above)
Comment (AlfonsoUceda) [reply test/automated/log_ignored/substitute.rb]:
```suggestion
    test "called with matching arguments" do
```

File: test/automated/log_ignored/substitute.rb [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+
+    test "called? returns false" do
+      assert substitute.called?
+    end
+
+    test "called_with? returns true for maatching arguments" do
+      assert substitute.called_with?(message, reason)
+    end
+
+    test "called_with? returns false for non matching reason" do
```
Comment (AlfonsoUceda) [inline test/automated/log_ignored/substitute.rb]: ?
```suggestion
    test "not called with other reason" do
```

File: test/automated/log_ignored/substitute.rb [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+      other_reason = Controls::Reason::Other.example
+
+      refute substitute.called_with?(message, other_reason)
+    end
+
+    test "called_with? returns false for non matching message" do
```
Comment (AlfonsoUceda) [inline test/automated/log_ignored/substitute.rb]: ?
```suggestion
    test "not called with other message" do
```

File: test/automated/log_ignored/substitute.rb (hunk 12 above)
Comment (AlfonsoUceda) [reply test/automated/log_ignored/substitute.rb]: Not, it isn't a copy, check again the tests :P

File: lib/log_ignored_message/controls.rb [hunk 19]
```diff
@@ -0,0 +1,6 @@
+require "eventide/postgres" # Requireed to define Read, which is used in messaging/controls
```
Comment (AlfonsoUceda) [inline lib/log_ignored_message/controls.rb]: I don't understand why this is needed. `Read` is a class defined inside `Messaging::Controls::Message`, this dependency shouldn't be used here and removed from this gem. Am I missing something?

File: test/automated.rb (hunk 2 above)
Comment (samstickland) [reply test/automated.rb]: @AlfonsoUceda Very odd. This came from the evt component skeleton.

File: lib/log_ignored_message/log_ignored.rb (hunk 7 above)
Comment (samstickland) [reply lib/log_ignored_message/log_ignored.rb]: I've removed this dependency for now. We can put it back in later if needed

File: test/automated/log_ignored/substitute.rb (hunk 12 above)
Comment (samstickland) [reply test/automated/log_ignored/substitute.rb]: No it's not. It's checking that it fails when the message is different, not reason. In order to implement this I had to investigate message equality - turns out it's based on attributes not ID.

File: lib/log_ignored_message/log_ignored.rb (hunk 13 above)
Comment (samstickland) [reply lib/log_ignored_message/log_ignored.rb]: I've kept it simple for now, and simply removed the namespace and made it `dependency LogIgnoredMessage, :log_ignore_message`

File: lib/log_ignored_message.rb (hunk 14 above)
Comment (samstickland) [reply lib/log_ignored_message.rb:1]: I decided not to do this, as I don't think including a gem should cause Object to get monkey patched. Instead I included the macro into LogIgnoredMessage.

File: lib/log_ignored_message/log_ignored.rb (hunk 15 above)
Comment (samstickland) [reply lib/log_ignored_message/log_ignored.rb]: I've removed this for now. I was trying to mimic the `include Log::Dependency` that logger has

File: lib/log_ignored_message/controls.rb (hunk 19 above)
Comment (samstickland) [reply lib/log_ignored_message/controls.rb]: When I include the controls for Message those controls reference the `Read` constant and the include fails with an exception.

Comment (AlfonsoUceda) [reply lib/log_ignored_message/controls.rb]: @samstickland You don't need this dependency you can remove it. You only need to do `require 'messaging'`

File: lib/log_ignored_message/log_ignored_message [hunk 20]
```diff
@@ -0,0 +1 @@
+/Users/sam/development/hubbado/hubbado-log_ignored_message/lib/log_ignored_message
```
Comment (arepnikov) [inline lib/log_ignored_message/log_ignored_message]: what is that? And what this file should do?

Comment (arepnikov) [reply lib/log_ignored_message/log_ignored_message]: maybe you missed file extension and `require`

**Review by arepnikov — APPROVED** [review]
Just one fix required, everything else looks ok

File: lib/log_ignored_message/log_ignored_message (hunk 20 above)
Comment (samstickland) [reply lib/log_ignored_message/log_ignored_message]: It's something that symlink-lib.sh added that shouldn't have had been comitted. Removing.
